harness/determined/pytorch/deepspeed/_deepspeed_trial.py

```diff
@@ -0,0 +1,881 @@
+        # careful to trigger its shutdown explicitly, to avoid hangs when the user is using
+        # multiprocessing-based parallelism for their dataloader.
+        #
+        # We create it before loading state because we don't want the training_iterator shuffling
+        # values after we load state.
+        self.training_iterator = (
+            iter(self.training_loader) if self.training_loader is not None else None
+        )
+        try:
+            # If a load path is provided load weights and restore the data location.
+            if self.env.latest_checkpoint is not None:
+                logging.info(f"Restoring trial from checkpoint {self.env.latest_checkpoint}")
+                with self.context._core.checkpointing.restore_path(
+                    self.env.latest_checkpoint
+                ) as load_path:
+                    self._load(pathlib.Path(load_path))
+
+            with self.prof:
+                for callback in self.callbacks.values():
+                    with self.prof.record_timing(
+                        f"callbacks.{callback.__class__.__name__}.on_training_start"
+                    ):
+                        callback.on_training_start()
+                self._run()
+
+        finally:
+            # Explicitly trigger the dataloader iterator shutdowns (which happens in __del__).
+            # See the rather long note in pytorch/torch/utils/data/dataloader.py.
+            if self.training_iterator is not None:
+                del self.training_iterator
+
+    def _run(self) -> None:
+        assert self.workloads is not None
+        for w, response_func in self.workloads:
+            try:
+                if w.kind == workload.Workload.Kind.RUN_STEP:
+                    action = "training"
+                    response = {
+                        "metrics": self._train_for_step(
+                            w.step_id,
+                            w.num_batches,
+                            w.total_batches_processed,
+                        ),
+                        "stop_requested": self.context.get_stop_requested(),
+                    }  # type: workload.Response
+
+                elif w.kind == workload.Workload.Kind.COMPUTE_VALIDATION_METRICS:
+                    action = "validation"
+                    response = {
+                        "metrics": self._compute_validation_metrics(),
+                        "stop_requested": self.context.get_stop_requested(),
+                    }
+
+                elif w.kind == workload.Workload.Kind.CHECKPOINT_MODEL:
+                    action = "checkpointing"
+                    # The checkpointing api would have been sufficient if the base_path for the
+                    # storage manager is guaranteed to be a shared file system.
+                    #
+                    # Since we can't guarantee that, we use the base storage_manager instead for
```

Comment:
can you make a jira ticket for updating this when we finish the checkpoint api updates?

Reply:
done